Re-implement public key ordering using underlying FFI functions

tcharding opened the pull request:
Re-base #309 for @dr-orlovsky on request by @Kixunil.

To do the rebase I just had to change instances of cfg_attr to use `v0_5_0` instead of `v0_4_1` e.g.,
```
    #[cfg_attr(not(rust_secp_no_symbol_renaming), link_name = "rustsecp256k1_v0_5_0_xonly_pubkey_cmp")]
```

And drop the changes to `src/schnorrsig.rs`, all these changes are covered by the changes in `key.rs` I believe.

Kixunil reviewed (commented):
Looks like https://github.com/rust-bitcoin/rust-secp256k1/pull/309#issuecomment-915597279 wasn't implemented.

> With cfg(fuzzing) now there is no implementation of Ord for PublicKey, which will cause downstream compliation failures during fuzzing.

> can you implement some sort of ordering for fuzzable public keys? You can even just auto-derive one.

tcharding commented on src/key.rs:
```diff
@@ -671,15 +671,25 @@ impl<'de> serde::Deserialize<'de> for PublicKey {
+            v if v < 0 => core::cmp::Ordering::Less,
+            v if v > 0 => core::cmp::Ordering::Greater,
+            _ => unreachable!()
+        }
```
ha! So it can :)

tcharding commented:
Changes in force push:
- Added `cmp` suggestion to the relevant patch (commit: `70a4441 Re-implementing PublicKey ordering using FFI`)
- Added `derive(Ord, PartialOrd)` to `PublicKey` using `cfg_attr` when fuzzing (added to last patch)

tcharding commented:
I have added/re-written the commit logs of the last two patches and added a `co-authored-by` tag, not to claim authorship but to accept blame if anything is wrong because these two patches have been non-trivially changed since @dr-orlovsky first posted them.

elichai commented:
I don't think we should put `#[cfg(not(fuzzing))]` on PartialOrd, instead we should probably add an impl to the fuzzing module (maybe something like just doing `==` between the arrays?)

Kixunil reviewed (approved):
ACK 13af51926a17f1f6b8a5797f6eb7ce0f1280b581

Kixunil commented:
@elichai care to explain why?

elichai commented:
> @elichai care to explain why?

I think any code using production rust-secp should compile under `fuzz` rust-secp, otherwise it makes fuzzing dependent projects much harder

I looked again now and saw `#[cfg_attr(feature = "fuzzing", derive(PartialOrd, Ord))]` which I missed the first time. 
So I guess this is fine, I still think we should try to "contain" the fuzz/production in the FFI modules and not let it leak into the whole codebase.

apoelstra reviewed (approved):
ACK 13af51926a17f1f6b8a5797f6eb7ce0f1280b581
